Assets/BossRoom/Scripts/Shared/Infrastructure/UpdateRunner.cs

```diff
@@ -0,0 +1,117 @@
+        /// Don't assume that onUpdate will be called in any particular order compared to other subscribers.
+        /// </summary>
+        public void Subscribe(Action<float> onUpdate, float period)
+        {
+            if (onUpdate == null)
+            {
+                return;
+            }
+
+            if (onUpdate.Target == null) // Detect a local function that cannot be Unsubscribed since it could go out of scope.
+            {
+                Debug.LogError("Can't subscribe to a local function that can go out of scope and can't be unsubscribed from");
+                return;
+            }
+
+            if (onUpdate.Method.ToString().Contains("<")) // Detect
+            {
+                Debug.LogError("Can't subscribe with an anonymous function that cannot be Unsubscribed, by checking for a character that can't exist in a declared method name.");
+                return;
+            }
+
+            foreach (var currSub in m_subscribers)
+            {
+                if (currSub.updateMethod.Equals(onUpdate))
+                {
+                    return;
+                }
+            }
+
+            m_subscribers.Add(new Subscriber(onUpdate, period));
+        }
+
+        /// <summary>Safe to call even if onUpdate was not previously Subscribed.</summary>
+        public void Unsubscribe(Action<float> onUpdate)
+        {
+            for (var sub = m_subscribers.Count - 1; sub >= 0; sub--)
+            {
+                if (m_subscribers[sub].updateMethod.Equals(onUpdate))
+                {
+                    m_subscribers.RemoveAt(sub);
+                }
+            }
+        }
+
+        private void Update()
+        {
+            OnUpdate(Time.deltaTime);
+        }
+
+        /// <summary>
+        /// Each frame, advance all subscribers. Any that have hit their period should then act, though if they take too long they could be removed.
+        /// </summary>
+        public void OnUpdate(float dt)
+        {
+            for (var subscriberIndex = m_subscribers.Count - 1; subscriberIndex >= 0; subscriberIndex--) // Iterate in reverse in case we need to remove something.
+            {
+                var subscriber = m_subscribers[subscriberIndex];
+                subscriber.periodCurrent += dt;
+
+                if (subscriber.periodCurrent > subscriber.period)
+                {
+                    var onUpdate = subscriber.updateMethod;
+
+                    if (onUpdate == null)
+                    {
+                        Remove(subscriberIndex, $"Did not Unsubscribe from UpdateSlow: {onUpdate.Target} : {onUpdate.Method}");
+                        continue;
+                    }
+
+                    onUpdate.Invoke(subscriber.periodCurrent);
+                    subscriber.periodCurrent = 0;
+                }
+            }
+
+            void Remove(int index, string msg)

```

Comment:
why do we need a method for this, why not just call removeAt directly at call site?

Reply:
Cleaned up